4/m_ipinfo_io.cpp

```diff
@@ -0,0 +1,189 @@
+        {
+            std::string url = "http://ipinfo.io/" + user->client_sa.addr() + "?token=" + apikey;
+            curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
+            res = curl_easy_perform(curl);
+            if (res != CURLE_OK)
+            {
+                std::lock_guard<std::mutex> lock(mtx);
+                ServerInstance->SNO.WriteGlobalSno('a', "IPInfo: Failed to get data for %s: %s", user->nick.c_str(), curl_easy_strerror(res));
+            }
+            else
+            {
+                ParseResponse(user, response);
+            }
+            curl_easy_cleanup(curl);
+        }
+        curl_global_cleanup();
+    }
+
+    void ParseResponse(LocalUser* user, const std::string& response)
+    {
+        rapidjson::Document document;
+        if (document.Parse(response.c_str()).HasParseError())
+        {
+            std::lock_guard<std::mutex> lock(mtx);
+            ServerInstance->SNO.WriteGlobalSno('a', "IPInfo: Failed to parse JSON for %s: %s", user->nick.c_str(), rapidjson::GetParseError_En(document.GetParseError()));
+            return;
+        }
+
+        std::string city = document.HasMember("city") ? document["city"].GetString() : "Unknown";
+        std::string region = document.HasMember("region") ? document["region"].GetString() : "Unknown";
+        std::string country = document.HasMember("country") ? document["country"].GetString() : "Unknown";
+        std::string org = document.HasMember("org") ? document["org"].GetString() : "Unknown";
+
+        std::string info = "City: " + city + ", Region: " + region + ", Country: " + country + ", Org: " + org;
+        cachedinfo.Set(user, info);
+
+        std::lock_guard<std::mutex> lock(mtx);
+        user->WriteNumeric(RPL_WHOISIPINFO, user->nick, "hes/her ip information: " + info);
+    }
+
+public:
+    IPInfoResolver(Module* Creator, LocalUser* user, StringExtItem& cache, const std::string& key)
+        : cachedinfo(cache), apikey(key), theiruuid(user->uuid)
+    {
+        worker = std::thread(&IPInfoResolver::DoRequest, this, user);
+    }
+
+    ~IPInfoResolver()
+    {
+        if (worker.joinable())
+        {
+            worker.join();
+        }
+    }
+};
+
+class ModuleIPInfo : public Module, public Stats::EventListener, public Whois::EventListener
+{
+private:
+    StringExtItem cachedinfo;
+    std::string apikey;
+
+public:
+    ModuleIPInfo()
+        : Module(VF_VENDOR, "Adds IPinfo.io information to WHOIS responses for opers, using a configured API key.")
+        , Stats::EventListener(this)
+        , Whois::EventListener(this)
+        , cachedinfo(this, "ipinfo", ExtensionType::USER)
+    {
+    }
+
+    void ReadConfig(ConfigStatus& status) override
+    {
+        auto& tag = ServerInstance->Config->ConfValue("ipinfo");
+        apikey = tag->getString("apikey", "", 1);
```

Comment:
You require `apikey` to be at least one character then but default it to an empty string anyway?